src/scales/scale.time.js

```diff
@@ -184,86 +170,182 @@ function parse(scale, input) {
+	let next = adapter.startOf(time, unit);
+	if (next < time) {
+		next = +adapter.add(next, 1, unit);
 	}
+	return next;
 }
 
-/**
- * Generates a maximum of `capacity` timestamps between min and max, rounded to the
- * `minor` unit using the given scale time `options`.
- * Important: this method can return ticks outside the min and max range, it's the
- * responsibility of the calling code to clamp values if needed.
- */
-function generate(scale, min, max, capacity) {
-	const adapter = scale._adapter;
-	const options = scale.options;
-	const timeOpts = options.time;
-	const minor = timeOpts.unit || determineUnitForAutoTicks(timeOpts.minUnit, min, max, capacity);
-	const stepSize = resolve([timeOpts.stepSize, timeOpts.unitStepSize, 1]);
-	const weekday = minor === 'week' ? timeOpts.isoWeekday : false;
-	const ticks = [];
+function getFirst(scale, minor, extendBounds) {
+	const {_adapter: adapter, min, _userMin} = scale;
+	const weekday = minor === 'week' ? scale.options.time.isoWeekday : false;
 	let first = min;
-	let time;
-
-	// For 'week' unit, handle the first day of week option
-	if (weekday) {
-		first = +adapter.startOf(first, 'isoWeek', weekday);
+	if (extendBounds && !_userMin) {
+		if (weekday) {
+			first = +adapter.startOf(first, 'isoWeek', weekday);
+		} else {
+			first = +adapter.startOf(min, minor);
+		}
 	}
+	return first;
+}
 
-	// Align first ticks on unit
-	first = +adapter.startOf(first, weekday ? 'day' : minor);
+function addTick(scale, tick, ticks, pixels) {
+	const checkPixels = scale.options.distribution === 'series' || scale.options.ticks.autoSkip;
 
-	// Prevent browser from freezing in case user options request millions of milliseconds
-	if (adapter.diff(max, min, minor) > 100000 * stepSize) {
-		throw min + ' and ' + max + ' are too far apart with stepSize of ' + stepSize + ' ' + minor;
-	}
+	if (checkPixels) {
+		const prev = pixels[pixels.length - 1];
 
-	for (time = first; time < max; time = +adapter.add(time, stepSize, minor)) {
-		ticks.push(time);
+		let pixel = scale.getPixelForValue(tick.value);
+		if (prev && Math.abs(pixel - prev.pixel) < scale._tickSpacing) {
+			if (tick.major && !prev.major) {
+				ticks.pop();
+				pixels.pop();
+			} else {
+				return false;
+			}
+		}
+		pixels.push({pixel, major: tick.major});
 	}
 
-	if (time === max || options.bounds === 'ticks') {
-		ticks.push(time);
+	return ticks.push(tick);
+}
+
+function generate(scale, units, stepSize) {
+	const {_adapter: adapter, options, max} = scale;
+	const tickOpts = options.ticks;
+	const axisLength = scale._length;
+	const extendBounds = options.bounds === 'ticks';
+	const pixels = [];
+	const minor = units[0];
+	const ticks = [];
+	let first = getFirst(scale, minor, extendBounds);
+	let majors = getMarjors(first, units, adapter);
+	let value = first;
+	let unit, major;
+
+	scale._tickSpacing = axisLength / tickOpts.maxTicksLimit || scale._resolveTickFontOptions(0).lineHeight + tickOpts.autoSkipPadding;
```

Comment:
I would think `lineHeight` would only work for vertical scales?

Reply:
Its used to estimate the maximum number of ticks that can fit (rotated 90 degrees)